### PR TITLE
typing: storage

### PR DESCRIPTION
**Pull request by aorenste**
This isn't a full typing of the file - it just fixes some uses of unbound 'T' (if you use a TypeVar as an output it also needs to be an input).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130670
* __->__ #130669

### CODE REVIEW

**Review by oulgen — APPROVED** [review]

File: torch/storage.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
 # mypy: allow-untyped-defs
+
+from __future__ import annotations
```
Comment (Skylion007) [inline torch/storage.py:3]: Why do we need to turn on future annotations here? Any particular reason?

File: torch/storage.py [hunk 2]
```diff
@@ -852,12 +862,15 @@ def untyped(self):
         _warn_typed_storage_removal()
         return self._untyped_storage
 
-    def _new_wrapped_storage(self, untyped_storage):
+    def _new_wrapped_storage(self, untyped_storage) -> Self:
         assert type(untyped_storage) == torch.UntypedStorage
 
         if type(self) == TypedStorage:
```
Comment (Skylion007) [inline torch/storage.py:868]:
```suggestion
        if type(self) is TypedStorage:
```

File: torch/storage.py (hunk 1 above)
Comment (aorenste) [reply torch/storage.py:3]: Sure - quoting works fine. But we could ask the opposite - what's the downside to turning it on?

File: torch/storage.py [hunk 3]
```diff
@@ -57,10 +59,14 @@ def nbytes(self) -> _int:
...
-    def type(self, dtype: _Optional[str] = None, non_blocking: _bool = False) -> T:  # type: ignore[type-var]
+    def type(
+        self, dtype: _Optional[str] = None, non_blocking: _bool = False
+    ) -> Union[_StorageBase, TypedStorage]:
         return _type(self, dtype, non_blocking)
 
-    def cuda(self, device=None, non_blocking=False) -> T:  # type: ignore[type-var, misc] # noqa: E704
+    def cuda(
+        self, device=None, non_blocking=False
+    ) -> Union[_StorageBase, TypedStorage]:
```
Comment (XuehaiPan) [inline torch/storage.py:69]: Add `from torch.types import Device`

```suggestion
    def cuda(
        self, device: Device = None, non_blocking: _bool = False
    ) -> Union[_StorageBase, TypedStorage]:
```

The argument should also be typed.

File: torch/storage.py [hunk 4]
```diff
@@ -1,11 +1,13 @@
 # mypy: allow-untyped-defs
```
Comment (XuehaiPan) [inline torch/storage.py:1]: We should type all methods and remove this comment.

Comment (aorenste) [reply torch/storage.py:1]: Totally agree - but that's a separate (and much larger) change (which I'm doing in a background task).  This is just removing some stuff that mypy actively complains about.

File: torch/storage.py (hunk 1 above)
Comment (Skylion007) [reply torch/storage.py]: New PEP is considering phasing it out.

Comment (aorenste) [reply torch/storage.py]: I assume you're talking about PEP749? From what I can tell it's just phasing out the import statement, not the deferred evaluation aspect that it provides - so sure we'd have to strip the import statement in something like 5 years from now - but the behavior should still be the same.

File: torch/storage.py (hunk 3 above)
Comment (aorenste) [reply torch/storage.py:69]: Is this correct (`Device`)? This function passes that argument to `torch.device` which is defined in Device.cpp (THPDevice_pynew) as taking `int64_t?` but `Device` is defined as `str | int | torch.device`.

**Review by Skylion007 — APPROVED** [review]

**Pull request closed without merging**